### ISSUE TITLE
DOMException to reason for failure commentary

### ISSUE THREAD

**Issue by james-d-elliott** [MasterKale/SimpleWebAuthn#183]
Hey just found this library while looking through suggested repos and was reading this comment: https://github.com/MasterKale/SimpleWebAuthn/issues/160#issuecomment-950263415

I wanted to discuss it but didn't want to derail that thread so figure this would be the ideal.

Unless I misunderstood the discussion or the specification, identifying the cause of why a failure occurred is technically possible as I read the spec. Providing (and of course this may not happen) browsers throw the correct DOMException depending on the exception and the task being performed, here is my notes as commented code as to what each exception means:

```ts
function getAttestationResultFromDOMException(exception: DOMException): AttestationResult {
    // Docs for this section:
    // https://w3c.github.io/webauthn/#sctn-op-make-cred
    switch (exception.name) {
        case "UnknownError":
            // § 6.3.2 Step 1 and Step 8.
            return AttestationResult.FailureSyntax;
        case "NotSupportedError":
            // § 6.3.2 Step 2.
            return AttestationResult.FailureSupport;
        case "InvalidStateError":
            // § 6.3.2 Step 3.
            return AttestationResult.FailureExcluded;
        case "NotAllowedError":
            // § 6.3.2 Step 3 and Step 6.
            return AttestationResult.FailureUserConsent;
        case "ConstraintError":
            // § 6.3.2 Step 4.
            return AttestationResult.FailureUserVerificationOrResidentKey;
        default:
            console.error(`Unhandled DOMException occurred during WebAuthN attestation: ${exception}`);
            return AttestationResult.FailureUnknown;
    }
}
```

```ts
function getAssertionResultFromDOMException(
    exception: DOMException,
    requestOptions: PublicKeyCredentialRequestOptions,
): AssertionResult {
    // Docs for this section:
    // https://w3c.github.io/webauthn/#sctn-op-get-assertion
    switch (exception.name) {
        case "UnknownError":
            // § 6.3.3 Step 1 and Step 12.
            return AssertionResult.FailureSyntax;
        case "NotAllowedError":
            // § 6.3.3 Step 6 and Step 7.
            return AssertionResult.FailureUserConsent;
        case "SecurityError":
            // § 10.1 and 10.2 Step 3.
            if (requestOptions.extensions?.appid !== undefined) {
                return AssertionResult.FailureU2FFacetID;
            } else {
                return AssertionResult.FailureUnknownSecurity;
            }
        default:
            console.error(`Unhandled DOMException occurred during WebAuthN assertion: ${exception}`);
            return AssertionResult.FailureUnknown;
    }
}
```

**Comment by MasterKale** [comment]
One of the first things I did after joining Duo was to discover all of the possible errors that could come out of `.create()` and `.get()` and what would throw them, and from that effort came a real ugly picture of how much understanding we could glean from any of the eight discrete errors defined in the spec. I don't have anything I can publicly share from that work yet (I plan on publishing this work to the Duo blog but no ETA), suffice to say `InvalidStateError` ended up being relatively straight forward to understand the user tried to re-register an authenticator matching one of the credentials in `excludeCredentials` during `.create()`.

Thank you for taking the time to write out that code. Looking over it I see that it's not all that unreasonable to try and derive what actually went wrong. Comparing your examples to the notes I left myself from my earlier research, I can actually see a path forward for making it possible to offer more insight into _most_ of the reasons an error was thrown.

Let me think on this and see how many other errors could also be intuited. It feels like a great way to make consuming WebAuthn simpler, and I'm all for that.

**Comment by james-d-elliott** [comment]
I'm rather interested in this, let me know if/when you get to post it.

I have noticed more consistency with browsers when it comes to Webauthn than compared to U2F which is nice, but I did suspect some inconsistencies and noticed some impossible to detect issues from the way the spec is worded itself, just found overall it was more consistent to finding the actual cause of a failure.

Also no worries about the code, it is actually from one of the projects I maintain. I wrote it in order to give user feedback during the process and to test if it were even possible with Webauthn.

**Comment by MasterKale** [comment]
@james-d-elliott Check out PR #184, I think this is the pattern I'll roll with since it won't change the current suggested pattern of wrapping `startRegistration()` in a `try / catch`.

I plan on adding similar logic for `.get()` to the PR before merging, at least the ones I know I can implement. There are some other errors raised that I don't think I can cleanly implement because of too much ambiguity (how can we distinguish a `NotAllowedError` in which the user fails to consent to create a new credential? See https://www.w3.org/TR/webauthn-2/#sctn-op-make-cred [Step 3]).

That and some extensions can raise identifiable errors, but extensions in general are poorly supported by me right now, so I can aim to identify those later if/when I find a way to abstract away the use of extensions.

**Comment by james-d-elliott** [comment]
Looks nice. I think the fact that it's at least potentially more informative is a good change. It means people implementing the client don't have to understand the spec too closely to implement some feedback to users.

**Comment by MasterKale** [comment]
> Looks nice. I think the fact that it's at least potentially more informative is a good change. It means people implementing the client don't have to understand the spec too closely to implement some feedback to users.

I realized too when I was waffling on whether to merge this in is that it would help provide standardized error messages across browsers that currently have varying degrees of insight into why the error occurred.

I'll cut a release with this new functionality soon, but I want to get PR #171 in before I do.

**Comment by james-d-elliott** [comment]
Yeah that'd be good. I have noticed some strange inconsistencies with mobile Safari. Not entirely sure if you've ran into this, but it appears it doesn't support appid via NFC potentially.

**Comment by MasterKale** [comment]
This functionality is now available in the newly-published **@simplewebauthn/browser@5.0.0** 🚀
